src/shared_memory_manager.cc

```diff
@@ -26,113 +26,59 @@
-{
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
-}
-#endif  // TRITON_ENABLE_GPU
-
-TRITONSERVER_Error*
-SharedMemoryManager::GetMemoryInfo(
-    const std::string& name, size_t offset, void** shm_mapped_addr,
-    TRITONSERVER_MemoryType* memory_type, int64_t* device_id)
-{
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
-}
-
-TRITONSERVER_Error*
-SharedMemoryManager::GetStatus(
-    const std::string& name, TRITONSERVER_MemoryType memory_type,
-    triton::common::TritonJson::Value* shm_status)
-{
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
-}
-
-TRITONSERVER_Error*
-SharedMemoryManager::Unregister(
-    const std::string& name, TRITONSERVER_MemoryType memory_type)
-{
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
-}
-
-TRITONSERVER_Error*
-SharedMemoryManager::UnregisterAll(TRITONSERVER_MemoryType memory_type)
-{
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
-}
+#include "common.h"
+#include "triton/common/logging.h"
 
-TRITONSERVER_Error*
-SharedMemoryManager::UnregisterHelper(
-    const std::string& name, TRITONSERVER_MemoryType memory_type)
-{
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
-}
-}}  // namespace triton::server
-#else
+// Not supporting shared memory for now
+#ifndef _WIN32
 #include <errno.h>
 #include <fcntl.h>
 #include <sys/mman.h>
 #include <unistd.h>
-
-#include "common.h"
-#include "triton/common/logging.h"
+#endif
 
 namespace triton { namespace server {
 
 namespace {
 
+#ifdef _WIN32
+const std::string
+GetWindowsError()
+{
+  LPSTR err_buffer = nullptr;
+  size_t size = FormatMessageA(
+      FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM |
+          FORMAT_MESSAGE_IGNORE_INSERTS,
+      NULL, GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+      (LPSTR)&err_buffer, 0, NULL);
+  std::string errstr(err_buffer, size);
+  std::string retString = std::to_string(GetLastError()) + " - " + errstr;
+  LocalFree(err_buffer);
+  return retString;
+}
+#endif
+
 TRITONSERVER_Error*
-OpenSharedMemoryRegion(const std::string& shm_key, int* shm_fd)
+OpenSharedMemoryRegion(const std::string& shm_key, void** shm_file)
 {
+#ifdef _WIN32
+  HANDLE* shm_handle = static_cast<HANDLE*>(shm_file);
```

Comment:
🧠

Reply:
To ponder the topic just a step further: I looked through the documentation of `static_cast` and expected there to be some sort of error if the cast is in someway "bad." There isn't, but probably because the underlying pointer could have been reassigned at some undetermined point in the code and then cast to a completely different type.

Reply:
Changed structure according to Guan's comments. We no longer make use of `static_cast` so our discussion no longer applies.